matrix_sdk_base/src/models/room.rs

```diff
@@ -341,126 +333,148 @@ impl Room {
+            return (false, HashMap::new());
+        }
+
         // Perform display name disambiguations, if necessary.
         let disambiguations =
-            self.disambiguation_updates(&leaving_member, MemberDirection::Exiting);
-        for (id, name) in disambiguations.into_iter() {
-            match name {
-                None => self.disambiguated_display_names.remove(&id),
-                Some(name) => self.disambiguated_display_names.insert(id, name),
-            };
-        }
+            self.disambiguation_updates(target_member, leaving_member.display_name.clone(), None);
 
-        if self.joined_members.contains_key(&leaving_member.user_id) {
-            self.joined_members.remove(&leaving_member.user_id);
-            true
-        } else if self.invited_members.contains_key(&leaving_member.user_id) {
-            self.invited_members.remove(&leaving_member.user_id);
-            true
-        } else {
-            false
+        debug!("remove_member: disambiguations: {:#?}", disambiguations);
+
+        for (id, is_ambiguous) in disambiguations.iter() {
+            self.get_member_mut(id).unwrap().display_name_ambiguous = *is_ambiguous;
         }
+
+        // TODO: factor this out to a method called `remove_member` and rename this method
+        // to something like `process_member_leaving_event`.
+        self.joined_members
+            .remove(target_member)
+            .or_else(|| self.invited_members.remove(target_member));
+
+        (true, disambiguations)
     }
 
-    /// Given a room `member`, return the list of members which have the same display name.
+    /// Check whether the user with the MXID `user_id` is joined or invited to the room.
     ///
-    /// The `inclusive` parameter controls whether the passed member should be included in the
-    /// list or not.
-    fn shares_displayname_with(&self, member: &RoomMember, inclusive: bool) -> Vec<UserId> {
+    /// Returns true if so, false otherwise.
+    pub fn member_is_tracked(&self, user_id: &UserId) -> bool {
+        self.invited_members.contains_key(&user_id) || self.joined_members.contains_key(&user_id)
+    }
+
+    /// Get a room member by user ID.
+    ///
+    /// If there is no such member, returns `None`.
+    pub fn get_member(&self, user_id: &UserId) -> Option<&RoomMember> {
+        self.joined_members
+            .get(user_id)
+            .or_else(|| self.invited_members.get(user_id))
+    }
+
+    /// Get a room member by user ID.
+    ///
+    /// If there is no such member, returns `None`.
+    pub fn get_member_mut(&mut self, user_id: &UserId) -> Option<&mut RoomMember> {
+        match self.joined_members.get_mut(user_id) {
+            None => self.invited_members.get_mut(user_id),
+            Some(m) => Some(m),
+        }
+    }
+
+    /// Given a display name, return the set of members which share it.
+    fn display_name_equivalence_class(&self, name: &str) -> HashSet<UserId> {
```

Comment:
Let's call this the `equivalence_set` since we're returning a set and the docstring mentions a set.

Reply:
I was going for [standard mathematical lingo](https://en.wikipedia.org/wiki/Equivalence_class) but this works as well.